### ISSUE TITLE
Dtype for VAE

### ISSUE THREAD

**Issue by NH1900** [luxiaolili/IDM-VTON-train#16]
Hey luxiaolili

Really thanks for sharing another training code here.

I have a question about dtype. For thesis official model training code https://github.com/yisol/IDM-VTON they did not have any mix_precision setup and forcely set VAE to fp16. 

Have you treid on official training code and did you met issue of mat1 and mat2 mismatch error? this is for the nn.Linear layer in Timesteps forward call will be posted. Since without mixed_precision layer weight and bias are float32 but input is fp16.

Also i saw in the training code 
madebyollin VAE is used which should be fixing the VAE on fp16 performance. However float32 was set to VAE. may want to check whether float 32 have big influnence on the final result?

Since I am currently facing the mask no converge issue (color or bright for mask area), was trying to eliminate reasons of dtype: attention layers weight training seems to be sensitive to float32 and fp16.

Any inputs would be much appreciated and thanks for your time in advance.

Best,
Kurisu

**Comment by luxiaolili** [comment]
> Hey luxiaolili
> 
> Really thanks for sharing another training code here.
> 
> I have a question about dtype. For thesis official model training code https://github.com/yisol/IDM-VTON they did not have any mix_precision setup and forcely set VAE to fp16.
> 
> Have you treid on official training code and did you met issue of mat1 and mat2 mismatch error? this is for the nn.Linear layer in Timesteps forward call will be posted. Since without mixed_precision layer weight and bias are float32 but input is fp16.
> 
> Also i saw in the training code madebyollin VAE is used which should be fixing the VAE on fp16 performance. However float32 was set to VAE. may want to check whether float 32 have big influnence on the final result?
> 
> Since I am currently facing the mask no converge issue (color or bright for mask area), was trying to eliminate reasons of dtype: attention layers weight training seems to be sensitive to float32 and fp16.
> 
> Any inputs would be much appreciated and thanks for your time in advance.
> 
> Best, Kurisu

Sorry, I never use the official model training code . sdxl vae for fp16 may cause the loss NaN. You can see https://huggingface.co/madebyollin/sdxl-vae-fp16-fix  , you will see sdxl vae decode failure in fp16

**Comment by NH1900** [comment]
> > Hey luxiaolili
> > Really thanks for sharing another training code here.
> > I have a question about dtype. For thesis official model training code https://github.com/yisol/IDM-VTON they did not have any mix_precision setup and forcely set VAE to fp16.
> > Have you treid on official training code and did you met issue of mat1 and mat2 mismatch error? this is for the nn.Linear layer in Timesteps forward call will be posted. Since without mixed_precision layer weight and bias are float32 but input is fp16.
> > Also i saw in the training code madebyollin VAE is used which should be fixing the VAE on fp16 performance. However float32 was set to VAE. may want to check whether float 32 have big influnence on the final result?
> > Since I am currently facing the mask no converge issue (color or bright for mask area), was trying to eliminate reasons of dtype: attention layers weight training seems to be sensitive to float32 and fp16.
> > Any inputs would be much appreciated and thanks for your time in advance.
> > Best, Kurisu
> 
> Sorry, I never use the official model training code . sdxl vae for fp16 may cause the loss NaN. You can see https://huggingface.co/madebyollin/sdxl-vae-fp16-fix , you will see sdxl vae decode failure in fp16

Gotcha thanks for the confirmation. I see your point. Just one minor follow up question, why did you use madebyollin instead of diffusers/stable-diffusion-xl-1.0-inpainting-0.1/vae since you already use float32 and the madebyollin is for fix when we pointing fp16 for VAE? I may miss some logic here please point it out if the question does not make sense.  

Thanks for your response for real.
